### PR TITLE
pending_merge

### PR DESCRIPTION
**Pull request by jhertz**
its a thing now!

there's a get and a put

still need UI

### CODE REVIEW

File: dao.py [hunk 1]
```diff
@@ -235,6 +237,14 @@ def get_players_with_similar_alias(self, alias):
         ret = self.players_col.find({'aliases': {'$in': similar_aliases}})
         return [Player.from_json(p) for p in ret]
 
+    def insert_pending_merge(self, the_merge):
+        return self.pending_merges_col.insert(the_merge.get_json_dict())
+
+    def get_pending_merge(self, merge_id):
+        print "entering get_pending_merge"
```
Comment (garsh0p) [inline dao.py]: don't forget to remove these prints

File: dao.py [hunk 2]
```diff
@@ -235,6 +237,14 @@ def get_players_with_similar_alias(self, alias):
...
+        info = self.pending_merges_col.find_one({'_id' : merge_id})
+        print "get_pending_merge info:", info
```
Comment (garsh0p) [inline dao.py]: this one too

File: model.py [hunk 3]
```diff
@@ -492,8 +494,41 @@ def from_json(cls, json_dict):
         if json_dict == None:
             return None
 
+        print "model from json", json_dict
```
Comment (garsh0p) [inline model.py]: remove print

File: model.py [hunk 4]
```diff
@@ -492,8 +494,41 @@ def from_json(cls, json_dict):
...
                 json_dict['_id'],
                 json_dict['admin_regions'],
                 json_dict['full_name'])
 
+
+class Merge(object):
+    def __init__(self, requester_user_id, base_player_obj_id, player_to_be_merged_obj_id, time, id=None):
+        self.requester_user_id = requester_user_id
+        self.base_player_obj_id = base_player_obj_id
+        self.player_to_be_merged_obj_id = player_to_be_merged_obj_id
```
Comment (garsh0p) [inline model.py]: base_player vs player_to_be_merged is still a little unclear to me. can you add a comment about which player actually disappears after this merge request is completed?

File: model.py [hunk 5]
```diff
@@ -492,8 +494,41 @@ def from_json(cls, json_dict):
...
+    def from_json(cls, json_dict):
+        if json_dict == None:
+            return None
+
+        return cls(
+                json_dict['requester_user_id'],
+                json_dict['base_player_obj_id'],
+                json_dict['player_to_be_merged_obj_id'],
+                json_dict['time'],
+                json_dict['_id']
```
Comment (garsh0p) [inline model.py]: This line should be like this: https://github.com/jhertz/garpr/blob/master/model.py#L330

File: model.py [hunk 6]
```diff
@@ -1,6 +1,8 @@
 from bson import json_util
 import json
 import trueskill
+from bson.objectid import ObjectId
```
Comment (garsh0p) [inline model.py]: This isn't used

File: server.py [hunk 7]
```diff
@@ -53,6 +53,10 @@
 tournament_put_parser.add_argument('matches', type=list)
 tournament_put_parser.add_argument('regions', type=list)
 
+merges_put_parser = reqparse.RequestParser()
+merges_put_parser.add_argument('base_player', type=str)
+merges_put_parser.add_argument('to_be_merged_player', type=str)
```
Comment (garsh0p) [inline server.py]: These are IDs correct? Can you suffix the variables with `_id`?

File: server.py [hunk 8]
```diff
@@ -550,6 +554,42 @@ def get(self):
 
         return return_dict
 
+class PendingMergesResource(restful.Resource):
+    def get(self):
+        pass
```
Comment (garsh0p) [inline server.py]: Did you mean to implement this or no?

File: server.py [hunk 9]
```diff
@@ -550,6 +554,42 @@ def get(self):
...
+
+    def put(self, region):
+        print "entering merge_put"
+        dao = Dao(region, mongo_client=mongo_client)
+        requesting_user = get_user_from_access_token(request.headers, dao)
+        if not is_user_admin_for_region(requesting_user, region): #wow, such auth
```
Comment (garsh0p) [inline server.py]: LOL good comment. I think what I wanted here "if the player is an admin for ANY region". can you change the check to `if request_user.admin_regions`?

Comment (garsh0p) [inline server.py]: or `if not request_user.admin_regions` rather since we're rejecting it here.

File: server.py [hunk 10]
```diff
@@ -550,6 +554,42 @@ def get(self):
...
+        # the above should validate that we have real objectIDs
+        # now lets validate that both of those players exist
+        if not dao.get_player_by_id(base_player):
+            return "base_player not found", 400
+        if not dao.get_player_by_id(to_be_merged_player):
+            return "to_be_merged_player not found", 400
+        #get curr time
+        now = datetime.now()
+        #create a new merge object
+        the_merge = Merge(requesting_user, base_player, to_be_merged_player, now)
```
Comment (garsh0p) [inline server.py]: you want `requesting_user.id` right, not the whole User object?

File: test/test_dao.py [hunk 11]
```diff
@@ -732,3 +732,24 @@ def test_get_all_users(self):
...
+        player_one = all_players[0]
+        player_two = all_players[1]
+        users = dao.get_all_users()
+        user = users[0]
+        now = datetime.now()
+        the_merge = Merge(user.id, player_one.id, player_two.id, now)
+
+        merge_id = dao.insert_pending_merge(the_merge)
+
+        self.assertTrue(merge_id)
```
Comment (garsh0p) [inline test/test_dao.py]: Can you assert that this is an instance of ObjectId?

File: test/test_dao.py [hunk 12]
```diff
@@ -732,3 +732,24 @@ def test_get_all_users(self):
...
+
+        the_merge_redux = dao.get_pending_merge(merge_id)
+
+        self.assertEqual(the_merge.base_player_obj_id, the_merge_redux.base_player_obj_id, msg=the_merge_redux)
```
Comment (garsh0p) [inline test/test_dao.py]: Can you assert the rest of the fields?

File: test/test_dao.py [hunk 13]
```diff
@@ -732,3 +732,24 @@ def test_get_all_users(self):
...
+
+
+
+        pass
```
Comment (garsh0p) [inline test/test_dao.py]: Delete this

File: test/test_dao.py [hunk 14]
```diff
@@ -732,3 +732,24 @@ def test_get_all_users(self):
         self.assertEquals(user.full_name, self.user_full_name_2)
         self.assertEquals(user.admin_regions, self.user_admin_regions_2)
 
+    def test_get_and_insert_pending_merge(self):
```
Comment (garsh0p) [inline test/test_dao.py]: Can you add a test to see that getting a merge with an ID that doesn't exist returns None?

File: server.py [hunk 15]
```diff
@@ -550,6 +554,42 @@ def get(self):
 
         return return_dict
 
+class PendingMergesResource(restful.Resource):
+    def get(self):
+        pass
+
+    def put(self, region):
```
Comment (garsh0p) [inline server.py]: This should actually be a POST not a PUT.

Comment (garsh0p) [inline server.py]: (fyi PUT usually specifies an ID or identifier. if you're creating a brand new object with a newly assigned ID, it's a POST)

File: test/test_server.py [hunk 16]
```diff
@@ -1260,3 +1260,71 @@ def test_put_player_permission_denied(self, mock_get_user_from_access_token):
         response = self.app.put('/texas/players/' + str(the_player.id), content_type='application/json')
         self.assertEquals(response.status_code, 403)
         self.assertEquals(response.data, '"Permission denied"')
+
```
Comment (garsh0p) [inline test/test_server.py]: Good tests! Thanks for this :)

**Comment by garsh0p** [comment]
Make sure you remove all those `prints`!

**Comment by ScootrNova** [comment]
happy to see some progress :+1:

File: dao.py (hunk 1 above)
Comment (jhertz) [inline dao.py]: done

File: dao.py (hunk 2 above)
Comment (jhertz) [inline dao.py]: done

File: model.py (hunk 3 above)
Comment (jhertz) [inline model.py]: done

File: model.py (hunk 4 above)
Comment (jhertz) [inline model.py]: done

File: server.py (hunk 7 above)
Comment (jhertz) [inline server.py]: done

File: server.py (hunk 8 above)
Comment (jhertz) [inline server.py]: i'm not sure. if the "review merges" stuff is going to be a real web-interface, then yes, we'd need this endpoint. OTOH, if you want to do it more as a script, then we don't really need this.

what do you think?

File: server.py (hunk 9 above)
Comment (jhertz) [inline server.py]: so we're letting admins for any region suggest merges? if so, I need to delete one of my test cases. just double-checking before i do that

File: server.py (hunk 10 above)
Comment (jhertz) [inline server.py]: yup, fixed

File: test/test_dao.py (hunk 11 above)
Comment (jhertz) [inline test/test_dao.py]: done

File: test/test_dao.py (hunk 12 above)
Comment (jhertz) [inline test/test_dao.py]: YES FINE GAR ILL DO IT CORRECTLY UGH YOURE THE WORST <3

pushing me to my potential *grumble*

File: test/test_dao.py (hunk 13 above)
Comment (jhertz) [inline test/test_dao.py]: yup

File: server.py (hunk 15 above)
Comment (jhertz) [inline server.py]: ugh, when i originally wrote this, i had it as POST, then i thought i was wrong, and switched it to PUT T_T

yeah its post now

File: test/test_server.py (hunk 16 above)
Comment (jhertz) [inline test/test_server.py]: :) its what i do


a lotttt more of those coming when we get there

File: test/test_dao.py (hunk 14 above)
Comment (jhertz) [inline test/test_dao.py]: done

File: test/test_dao.py (hunk 12 above)
Comment (jhertz) [inline test/test_dao.py]: hey, TDD is the best, i found a bug, one of the ids is none, gonna have to fix that

File: server.py [hunk 17]
```diff
@@ -550,6 +554,41 @@ def get(self):
...
 
+class PendingMergesResource(restful.Resource):
+    def get(self):
+        #TODO: decide if we want to implement this
+        pass
+
+    def post(self, region):
+        dao = Dao(region, mongo_client=mongo_client)
+        requesting_user = get_user_from_access_token(request.headers, dao)
+        if not is_user_admin_for_region(requesting_user, region): #wow, such auth
```
Comment (garsh0p) [inline server.py]: sorry i forgot to reply to your last comment, but can you change this to `if not request_user.admin_regions`?

File: test/test_dao.py [hunk 18]
```diff
@@ -732,3 +732,31 @@ def test_get_all_users(self):
...
+        self.assertIsInstance(merge_id, ObjectId)
+        self.assertEqual(merge_id, orig_id)
+
+        the_merge_redux = dao.get_pending_merge(merge_id)
+
+        self.assertEqual(the_merge.base_player_obj_id, the_merge_redux.base_player_obj_id, msg=the_merge_redux)
+        self.assertEqual(the_merge.player_to_be_merged_obj_id, the_merge_redux.player_to_be_merged_obj_id)
+        self.assertEqual(the_merge.id, the_merge_redux.id)
+        self.assertEqual(the_merge.requester_user_id, the_merge_redux.requester_user_id)
+        self.assertTrue(abs(the_merge.time - the_merge_redux.time).total_seconds() < 1) # changed to account for mongo driver losing sub-second accuracy
```
Comment (garsh0p) [inline test/test_dao.py]: nit: put the comment on the line above, this line is too long

Comment (jhertz) [inline test/test_dao.py]: done!

File: server.py (hunk 17 above)
Comment (jhertz) [inline server.py]: done!

**Comment by garsh0p** [comment]
I fixed `test_server.test_post_not_admin` to pass after your changes @jhertz and merged that along with your code.
